xarray/core/dataset.py

```diff
@@ -1320,7 +1320,7 @@ def _construct_dataarray(self, name: Hashable) -> "DataArray":
         needed_dims = set(variable.dims)
 
         coords: Dict[Hashable, Variable] = {}
-        for k in self.coords:
+        for k in self._coord_names:
```

Comment:
The test failures are an ordering issue. 

However the "iterate over `self.variables`" trick doesn't work (e.g. #4409) too well for this microbenchmark because it has 2000 data variables and only 1 coordinate variable. It's still a 2x speedup (2.6s → 1.3s) so could be worth doing anyway.

Reply:
I don't think 2000:1 is a particularly realistic ratio :). I do think iterating over `self._variables` first is probably the right answer, for the ordering reason. The other option is to switch to some sort of ordered set for `_coords` (but Python doesn't have any built-in data structure for that).